File: app/src/main/java/de/saschahlusiak/frupic/gallery/GalleryAdapter.java
package de.saschahlusiak.frupic.gallery;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.util.Date;

import com.davemorrissey.labs.subscaleview.ImageSource;
import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView;

import de.saschahlusiak.frupic.R;
import de.saschahlusiak.frupic.cache.FileCacheUtils;
import de.saschahlusiak.frupic.model.Frupic;
import de.saschahlusiak.frupic.model.FrupicFactory;
import de.saschahlusiak.frupic.services.FetchFrupicJob;
import de.saschahlusiak.frupic.services.Job;
import de.saschahlusiak.frupic.services.Job.OnJobListener;
import de.saschahlusiak.frupic.services.Job.Priority;
import de.saschahlusiak.frupic.services.JobManager;
import android.database.Cursor;
import android.net.Uri;
import androidx.viewpager.widget.PagerAdapter;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.TextView;

public class GalleryAdapter extends PagerAdapter implements OnJobListener {
	private final static String tag = GalleryAdapter.class.getSimpleName();

	private GalleryActivity context;
	private Cursor cursor;
	private boolean showAnimations;
	private FrupicFactory factory;
	private JobManager jobManager;
	
	static class ViewHolder {
		public Frupic frupic;
		public Job job;
		public View view;
		
		public ViewGroup progressLayout;
		public ProgressBar progress;
		public TextView progressText;
	}

	public GalleryAdapter(GalleryActivity context, boolean showAnimations) {
		this.context = context;
		this.showAnimations = showAnimations;
		this.factory = new FrupicFactory(context);
	}
	
	public void setJobManager(JobManager jobManager) {
		if (jobManager == null && this.jobManager != null) {
			this.jobManager.removeAllJobListener(this);
		}
		this.jobManager = jobManager;
		
		notifyDataSetChanged();
	}

	public void setCursor(Cursor cursor) {
		this.cursor = cursor;
		notifyDataSetChanged();
	}
	
	private boolean showFrupic(View view, Frupic frupic) {
		GifMovieView v = (GifMovieView)view.findViewById(R.id.videoView);
		SubsamplingScaleImageView i = (SubsamplingScaleImageView)view.findViewById(R.id.imageView);

		if (showAnimations && frupic.isAnimated()) {
			v.setVisibility(View.VISIBLE);
			i.setVisibility(View.GONE);
			String filename = new FileCacheUtils(context).getFileName(frupic);
            InputStream stream = null;
			try {
				/* Movie calls reset() which the InputStream must support.
				 * 
				 * FileInputStream does NOT
				 * BufferedInputStream does, but somehow we need to call mark(x) first, with x>0. WTF?
				 * ByteArrayInputStream does
				 * 
				 * I do not trust BufferedInputStream and because it will probably map the whole file anyway,
				 * I can just save it in a byte array.
				 */
				stream = new FileInputStream(filename);
				
//				stream = new BufferedInputStream(stream);
//				stream.mark(1);
				
				ByteArrayOutputStream bos = new ByteArrayOutputStream();
				byte buf[] = new byte[4096];
				while (stream.read(buf) > 0) {
					bos.write(buf);
				}
				bos.flush();
				stream.close();
				stream = new ByteArrayInputStream(bos.toByteArray());
				
				v.setStream(stream);
				return true;
			} catch (OutOfMemoryError e2) {
				stream = null;
				System.gc();
				Log.e("OutOfMemoryError", "trying to load gif animation as Bitmap instead");
				/* fall-through to load Bitmap instead*/
			} catch (FileNotFoundException e1) {
				return false;
			} catch (Exception e1) {
				e1.printStackTrace();
				return false;
			}			
		}
		
		/* fall-through, if loading animation failed */
		File file = factory.getCacheFile(frupic);
		Uri uri = null;
		if (file.exists()) {
			uri = Uri.fromFile(file);
			file.setLastModified(new Date().getTime());
			
			i.setVisibility(View.VISIBLE);
			v.setVisibility(View.GONE);
			i.setImage(ImageSource.uri(uri));
			
			return true;
		}
		return false;
	}

	@Override
	public Object instantiateItem(ViewGroup container, int position) {
		Log.w(tag, "instantiateItem(" + position + ")");
		cursor.moveToPosition(position);

		Frupic frupic = new Frupic(cursor);

		View view = LayoutInflater.from(context).inflate(R.layout.gallery_item, container, false);
		SubsamplingScaleImageView i = (SubsamplingScaleImageView)view.findViewById(R.id.imageView);
		GifMovieView v = (GifMovieView)view.findViewById(R.id.videoView);
		
		i.setOrientation(SubsamplingScaleImageView.ORIENTATION_USE_EXIF);
		
		// TODO: context menu on image view seems to be broken
		context.registerForContextMenu(view);
		context.registerForContextMenu(i);
		context.registerForContextMenu(v);
		
		OnClickListener l = new OnClickListener() {
			@Override
			public void onClick(View v) {
				context.toggleControls();
			}
		};
		
		i.setOnClickListener(l);
		v.setOnClickListener(l);
		view.setOnClickListener(l);
		
		final ViewHolder holder = new ViewHolder();
		holder.view = view;
		holder.progressLayout = (ViewGroup) view.findViewById(R.id.progressLayout);
		
		holder.progress = (ProgressBar)view.findViewById(R.id.progressBar);
		holder.progress.setIndeterminate(false);
		holder.progress.setMax(100);
		holder.progress.setProgress(0);
		holder.progressText = (TextView)view.findViewById(R.id.progress);
		
		final ImageButton stopButton = (ImageButton)holder.view.findViewById(R.id.stopButton);
		
		stopButton.setImageResource(android.R.drawable.ic_menu_close_clear_cancel);
		stopButton.setVisibility(View.VISIBLE);
		stopButton.setOnClickListener(new OnClickListener() {
			
			@Override
			public void onClick(View v) {
				if (holder.job.isCancelled()) {
					jobManager.post(holder.job, Priority.PRIORITY_HIGH);
				} else {
					holder.job.cancel();
					stopButton.setImageResource(android.R.drawable.ic_menu_revert);
				}
			}
		});
		
		holder.frupic = frupic;
		
		frupic.tag = holder;
		view.setTag(holder);

		if (!showFrupic(view, frupic)) {
			i.setVisibility(View.GONE);
			v.setVisibility(View.GONE);
			
			holder.progress.setVisibility(View.VISIBLE);
			holder.progressText.setVisibility(View.VISIBLE);
			holder.progressText.setText(R.string.waiting_to_start);
		
			if (jobManager != null) {
				holder.job = jobManager.getFetchJob(frupic, factory);
				holder.job.addJobListener(this);
				
				jobManager.post(holder.job, Priority.PRIORITY_HIGH);
			} else {
				Log.d(tag, "JobManager not available yet");
				holder.view = null;
			}	
		} else {
			view.findViewById(R.id.progressLayout).setVisibility(View.GONE);
		}
		
		container.addView(view);
		return holder;
	}
	
	@Override
	public void destroyItem(ViewGroup container, int position, Object object) {
		Log.w(tag, "destroyItem(" + position + ")");
		
		ViewHolder holder = (ViewHolder)object;
		holder.frupic.tag = null;
		if (holder.job != null)
			holder.job.cancel();
		
		container.removeView(holder.view);
	}
	
	@Override
	public int getItemPosition(Object object) {
		ViewHolder holder = (ViewHolder)object;
		if (holder.view == null)
			return POSITION_NONE;
		return POSITION_UNCHANGED;
	}

	@Override
	public int getCount() {
		if (cursor == null)
			return 0;
		return cursor.getCount();
	}

	@Override
	public boolean isViewFromObject(View view, Object object) {
		return view == ((ViewHolder)object).view;
	}

	@Override
	public void OnJobStarted(final Job job) {
		Frupic frupic = ((FetchFrupicJob)job).getFrupic();
		final ViewHolder holder = (ViewHolder) frupic.tag;
		
		
	}

	@Override
	public void OnJobProgress(Job job, final int progress, final int max) {
		Frupic frupic = ((FetchFrupicJob)job).getFrupic();
		final ViewHolder holder = (ViewHolder) frupic.tag;
	
		if (holder == null)
			return;

		holder.view.post(new Runnable() {
			@Override
			public void run() {
				holder.progress.setProgress((100 * progress) / max);
				holder.progressText.setText(String.format("%dkb / %dkb (%d%%)", progress / 1024, max / 1024, (max > 0) ? progress * 100 / max : 0));
			}
		});
	}

	@Override
	public void OnJobDone(Job job) {
		Frupic frupic = ((FetchFrupicJob)job).getFrupic();
		ViewHolder holder = (ViewHolder) frupic.tag;
		
		Log.d(tag, "OnJobDone(" + frupic.id + ")");
		
		if (holder == null)
			return;
		
		if (job.isCancelled()) {
			holder.progressText.setText(R.string.cancelled);
		} else {
			holder.job.removeJobListener(this);
			holder.job = null;
			SubsamplingScaleImageView i = (SubsamplingScaleImageView) holder.view.findViewById(R.id.imageView);
			GifMovieView v = (GifMovieView)holder.view.findViewById(R.id.videoView);
			
			holder.progressLayout.setVisibility(View.GONE);
			
			if (!showFrupic(holder.view, frupic)) {
				i.setVisibility(View.VISIBLE);
				v.setVisibility(View.GONE);
				i.setImage(ImageSource.resource(R.drawable.broken_frupic));
			}
		}
	}
}

File: app/src/main/java/de/saschahlusiak/frupic/gallery/GalleryAdapter.kt
package de.saschahlusiak.frupic.gallery

import android.net.Uri
import android.util.Log
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.Observer
import androidx.lifecycle.asLiveData
import androidx.viewpager.widget.PagerAdapter
import com.davemorrissey.labs.subscaleview.ImageSource
import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView
import de.saschahlusiak.frupic.R
import de.saschahlusiak.frupic.app.DownloadJob
import de.saschahlusiak.frupic.app.FrupicDownloadManager
import de.saschahlusiak.frupic.app.FrupicStorage
import de.saschahlusiak.frupic.app.JobStatus
import de.saschahlusiak.frupic.databinding.GalleryItemBinding
import de.saschahlusiak.frupic.model.Frupic
import java.io.*
import java.util.*

class GalleryAdapter(
    private val activity: GalleryActivity,
    private val showAnimations: Boolean,
    private val storage: FrupicStorage,
    private val downloadManager: FrupicDownloadManager
) : PagerAdapter() {
    private var list: List<Frupic> = emptyList()

    inner class ViewHolder(
        val binding: GalleryItemBinding,
        val frupic: Frupic
    ) {
        private val progressLayout = binding.progressLayout
        private val stopButton = binding.stopButton
        private val progressBar = binding.progressBar
        private val progressText = binding.progress
        private val image = binding.imageView
        private val video = binding.videoView

        init {
            progressBar.apply {
                isIndeterminate = false
                max = 100
                progress = 0
            }

            stopButton.setOnClickListener {
                stopDownloadClick()
            }

            newFrupic(frupic)
        }

        private fun newFrupic(frupic: Frupic) {
            if (showFrupic(binding.root, frupic)) {
                // loaded and shown, otherwise download
                progressLayout.visibility = View.GONE
            } else {
                progressLayout.visibility = View.VISIBLE
                image.visibility = View.INVISIBLE
                video.visibility = View.INVISIBLE

                val job = getDownloadJob(frupic)
                job.status.asLiveData().observe(activity, Observer { result ->
                    when (result) {
                        is JobStatus.InProgress -> {
                            progressBar.progress = 100 * result.progress / result.max
                            progressText.text = String.format(
                                "%dkb / %dkb (%d%%)",
                                result.progress / 1024,
                                result.max / 1024,
                                if (result.max > 0) result.progress * 100 / result.max else 0
                            )
                        }

                        JobStatus.Cancelled -> {
                            progressText.setText(R.string.cancelled)
                            stopButton.setImageResource(android.R.drawable.ic_menu_revert)
                        }

                        is JobStatus.Success,
                            JobStatus.Failed -> {
                            progressLayout.visibility = View.GONE
                            if (!showFrupic(binding.root, frupic)) {
                                image.visibility = View.VISIBLE
                                video.visibility = View.GONE
                                image.setImage(ImageSource.resource(R.drawable.broken_frupic))
                            }
                        }

                        JobStatus.Scheduled -> {
                            // FIXME
                        }
                    }
                })
            }
        }

        private fun getDownloadJob(frupic: Frupic): DownloadJob {
            val existing = downloadManager.getJobOrNull(frupic)
            if (existing != null) return existing

            progressText.setText(R.string.waiting_to_start)
            // TODO: replace with vector drawable
            stopButton.setImageResource(android.R.drawable.ic_menu_close_clear_cancel)

            return DownloadJob(frupic).also {
                downloadManager.enqueue(it)
            }
        }

        private fun stopDownloadClick() {
            if (downloadManager.cancel(frupic)) {
                // nothing
            } else {
                newFrupic(frupic)
            }
        }
    }

    fun setList(list: List<Frupic>) {
        this.list = list
        notifyDataSetChanged()
    }

    private fun showFrupic(view: View, frupic: Frupic): Boolean {
        val v: GifMovieView = view.findViewById(R.id.videoView)
        val i: SubsamplingScaleImageView = view.findViewById(R.id.imageView)
        if (showAnimations && frupic.isAnimated) {
            v.visibility = View.VISIBLE
            i.visibility = View.GONE
            val file = storage.getFile(frupic)
            var stream: InputStream
            try {
                /* Movie calls reset() which the InputStream must support.
				 *
				 * FileInputStream does NOT
				 * BufferedInputStream does, but somehow we need to call mark(x) first, with x>0. WTF?
				 * ByteArrayInputStream does
				 *
				 * I do not trust BufferedInputStream and because it will probably map the whole file anyway,
				 * I can just save it in a byte array.
				 */
                stream = FileInputStream(file)

//				stream = new BufferedInputStream(stream);
//				stream.mark(1);
                val bos = ByteArrayOutputStream()
                val buf = ByteArray(4096)
                while (stream.read(buf) > 0) {
                    bos.write(buf)
                }
                bos.flush()
                stream.close()
                stream = ByteArrayInputStream(bos.toByteArray())
                v.setStream(stream)
                return true
            } catch (e2: OutOfMemoryError) {
                System.gc()
                Log.e("OutOfMemoryError", "trying to load gif animation as Bitmap instead")
                /* fall-through to load Bitmap instead*/
            } catch (e1: FileNotFoundException) {
                return false
            } catch (e1: Exception) {
                e1.printStackTrace()
                return false
            }
        }

        /* fall-through, if loading animation failed */
        val file = storage.getFile(frupic)

        if (file.exists()) {
            val uri = Uri.fromFile(file)
            file.setLastModified(Date().time)
            i.visibility = View.VISIBLE
            v.visibility = View.GONE
            i.setImage(ImageSource.uri(uri))
            return true
        }

        return false
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val frupic = list[position]

        Log.w(tag, "instantiateItem($position)")
        val binding = GalleryItemBinding.inflate(activity.layoutInflater, container, false)
        binding.imageView.orientation = SubsamplingScaleImageView.ORIENTATION_USE_EXIF

        val l = View.OnClickListener { activity.toggleControls() }
        binding.imageView.setOnClickListener(l)
        binding.videoView.setOnClickListener(l)
        binding.root.setOnClickListener(l)
        val holder = ViewHolder(binding, frupic)

        binding.root.tag = holder

        container.addView(binding.root)
        return holder
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        Log.w(tag, "destroyItem($position)")
        val holder = `object` as ViewHolder
        downloadManager.cancel(holder.frupic)
        container.removeView(holder.binding.root)
    }

    override fun getItemPosition(`object`: Any): Int {
        val holder = `object` as ViewHolder
        val frupic = holder.frupic
        val found = list.firstOrNull {
            it.id == frupic.id
        } ?: POSITION_NONE
        return list.indexOf(found)
    }

    override fun getCount(): Int {
        return list.size
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view === (`object` as ViewHolder).binding.root
    }

    companion object {
        private val tag = GalleryAdapter::class.java.simpleName
    }
}